### PR TITLE
Add tests for `matchers.py`

### PR DESCRIPTION
**Pull request by HiromuHota**
## Description of the problems or issues

**Is your pull request related to a problem? Please describe.**

Towards #380.

**Does your pull request fix any issue.**

Towards #380.

## Description of the proposed changes

This PR adds more tests.

## Test plan

This PR adds more tests.

## Checklist

* [ ] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.
* [ ] All new and existing tests passed.
* [ ] I have updated the CHANGELOG.rst accordingly.

### CODE REVIEW

File: tests/candidates/test_matchers.py [hunk 1]
```diff
@@ -173,3 +196,231 @@ def test_inverse(doc_setup):
...
+    assert set(tc.get_span() for tc in matcher.apply(space.apply(doc))) == {"This is"}
+
+    # Test with permutations=False
+    matcher = Concat(matcher1, matcher0, permutations=False)
+    assert set(matcher.apply(space.apply(doc))) == set()
+    # Test with permutations=True
+    matcher = Concat(matcher1, matcher0, permutations=True)
+    assert set(tc.get_span() for tc in matcher.apply(space.apply(doc))) == {"This is"}
+
+    # TODO: Add a test for ignore_sep=False
```
Comment (SenWu) [inline tests/candidates/test_matchers.py:244]: Are we going to have this test? Thanks!

File: tests/candidates/test_matchers.py [hunk 2]
```diff
@@ -173,3 +196,231 @@ def test_inverse(doc_setup):
...
+        "O",
+        "O",
+        "O",
+        "GPE",
+        "O",
+        "DATE",
+        "O",
+    ]
+    doc.sentences[1].ner_tags = ["O", "O", "O", "O", "O", "ORG", "O"]
+    # TODO: replace "NUMBER" with "CARDINAL" (#473)
```
Comment (SenWu) [inline tests/candidates/test_matchers.py:361]: Let's make this in another PR.
Maybe check this depends on SpaCy version?

Comment (HiromuHota) [reply tests/candidates/test_matchers.py:361]: I'd fix this in another PR (which will fix #473).

> Maybe check this depends on SpaCy version?

What do you mean by this?

File: tests/candidates/test_matchers.py (hunk 1 above)
Comment (HiromuHota) [reply tests/candidates/test_matchers.py:244]: I was not going to as I don't know how to use `ignore_sep=False`.
But let me see the code and try to implement it!

Comment (HiromuHota) [reply tests/candidates/test_matchers.py:244]: This option dates back to https://github.com/snorkel-team/snorkel/issues/262 and https://github.com/snorkel-team/snorkel-extraction/commit/4eb2907f10fb8652fa16296e625ad1bc47e31b9e.

Comment (HiromuHota) [reply tests/candidates/test_matchers.py:244]: I think I found a bug around options `sep` and `ignore_sep`, which is not easy to fix.
Let's keep it TODO here instead of writing tests that just pass and fool us.
